cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -0,0 +1,291 @@
+  // RLC = 101
+  CheckIfElseOutputAAA(type, "[true, true, true, false]", "[1, null, 3, 4]",
+                       "[5, 6, 7, 8]", "[1, null, 3, 8]", false);
+  // RLC = 100
+  CheckIfElseOutputAAA(type, "[true, true, null, false]", "[1, null, 3, 4]",
+                       "[5, 6, 7, 8]", "[1, null, null, 8]", false);
+  // RLC = 011
+  CheckIfElseOutputAAA(type, "[true, true, true, false]", "[1, 2, 3, 4]",
+                       "[5, 6, 7, null]", "[1, 2, 3, null]", false);
+  // RLC = 010
+  CheckIfElseOutputAAA(type, "[null, true, true, false]", "[1, 2, 3, 4]",
+                       "[5, 6, 7, null]", "[null, 2, 3, null]", false);
+  // RLC = 001
+  CheckIfElseOutputAAA(type, "[true, true, true, false]", "[1, 2, null, null]",
+                       "[null, 6, 7, null]", "[1, 2, null, null]", false);
+  // RLC = 000
+  CheckIfElseOutputAAA(type, "[null, true, true, false]", "[1, 2, null, null]",
+                       "[null, 6, 7, null]", "[null, 2, null, null]", false);
+
+  using ArrayType = typename TypeTraits<TypeParam>::ArrayType;
+  random::RandomArrayGenerator rand(/*seed=*/0);
+  int64_t len = 1000;
+  auto cond = std::static_pointer_cast<BooleanArray>(
+      rand.ArrayOf(boolean(), len, /*null_probability=*/0.01));
+  auto left = std::static_pointer_cast<ArrayType>(
+      rand.ArrayOf(type, len, /*null_probability=*/0.01));
+  auto right = std::static_pointer_cast<ArrayType>(
+      rand.ArrayOf(type, len, /*null_probability=*/0.01));
+
+  typename TypeTraits<TypeParam>::BuilderType builder;
+
+  for (int64_t i = 0; i < len; ++i) {
+    if (!cond->IsValid(i) || (cond->Value(i) && !left->IsValid(i)) ||
+        (!cond->Value(i) && !right->IsValid(i))) {
+      ASSERT_OK(builder.AppendNull());
+      continue;
+    }
+
+    if (cond->Value(i)) {
+      ASSERT_OK(builder.Append(left->Value(i)));
+    } else {
+      ASSERT_OK(builder.Append(right->Value(i)));
+    }
+  }
+  ASSERT_OK_AND_ASSIGN(auto expected_data, builder.Finish());
+
+  CheckIfElseOutputArray(cond, left, right, expected_data, false);
+
+  // -------- Cond - Array, Left- Array, Right - Scalar ---------
+
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Scalar> valid_scalar, MakeScalar(type, 100));
+  std::shared_ptr<Scalar> null_scalar = MakeNullScalar(type);
+
+  // empty
+  CheckIfElseOutputAAS(type, "[]", "[]", valid_scalar, "[]");
+
+  // RLC = 111
+  CheckIfElseOutputAAS(type, "[true, true, true, false]", "[1, 2, 3, 4]", valid_scalar,
+                       "[1, 2, 3, 100]");
+  // RLC = 110
+  CheckIfElseOutputAAS(type, "[true, true, null, false]", "[1, 2, 3, 4]", valid_scalar,
+                       "[1, 2, null, 100]", false);
+  // RLC = 101
+  CheckIfElseOutputAAS(type, "[true, true, true, false]", "[1, null, 3, 4]", valid_scalar,
+                       "[1, null, 3, 100]", false);
+  // RLC = 100
+  CheckIfElseOutputAAS(type, "[true, true, null, false]", "[1, null, 3, 4]", valid_scalar,
+                       "[1, null, null, 100]", false);
```

Comment:
Ok, I see from above that it's about the validity of cond/left/right. IMO if you're going to abbreviate it, at least always write it as C/L/R since that's the order of the arguments.

Reply:
Ah, and these probably can't be combined as you want to test null handling for different cases as well. Ok, got it.